patchwork-events-entity/src/main/java/net/patchworkmc/mixin/event/entity/MixinItemEntity.java

```diff
@@ -0,0 +1,132 @@
+
+@Mixin(ItemEntity.class)
+public abstract class MixinItemEntity {
+	@Shadow
+	private int pickupDelay;
+
+	@Shadow
+	public abstract ItemStack getStack();
+
+	@Unique
+	public int lifespan = 6000;
+
+	// TODO -> Forge has exposed 'lifespan' (hardcoded to 6000) through IForgeItemStack/IForgeItem#getEntityLifespan
+	/*
+		@Inject(method = "<init>(Lnet/minecraft/world/World;DDDLnet/minecraft/item/ItemStack;)V", at = @At("RETURN"))
+		private void modifyLifespan(World world, double x, double y, double z, ItemStack stack, CallbackInfo ci) {
+			this.lifespan = stack.getItem() == null ? 6000 : ((IForgeItemStack) stack).getEntityLifespan(world);
+		}
+	 */
+
+	// TODO -> Forge has a callback at IForgeItemStack/IForgeItem#onEntityItemUpdate
+	/*
+		@Inject(method = "tick", at = @At("HEAD"), cancellable = true)
+		private void onUpdate(CallbackInfo ci) {
+			if ((IForgeItemStack) stack).onEntityItemUpdate((ItemEntity) (Object) this) {
+				ci.cancel();
+			}
+		}
+	 */
+
+	@ModifyConstant(method = "tick", constant = @Constant(intValue = 6000))
+	private int onTickCheckLifespan(int original) {
+		return lifespan;
+	}
+
+	@Redirect(method = "tick", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/ItemEntity;remove()V"))
+	private void fireOnItemExpire(ItemEntity item) {
+		int hook = EntityEvents.onItemExpire(item, getStack());
+
+		if (hook < 0) {
+			item.remove();
+		} else {
+			this.lifespan += hook;
+		}
+	}
+
+	@Inject(method = "writeCustomDataToTag", at = @At("HEAD"))
+	private void addLifespanToTag(CompoundTag tag, CallbackInfo ci) {
+		tag.putInt("Lifespan", lifespan);
+	}
+
+	@Inject(method = "readCustomDataFromTag", at = @At("HEAD"))
+	private void readLifespanInTag(CompoundTag tag, CallbackInfo ci) {
+		if (tag.contains("Lifespan")) {
+			this.lifespan = tag.getInt("Lifespan");
+		}
+	}
+
+	@Redirect(method = "onPlayerCollision", at = @At(value = "FIELD", opcode = Opcodes.GETFIELD, target = "Lnet/minecraft/world/World;isClient:Z"))
+	private boolean checkPickupDelay(World world) {
+		return world.isClient && this.pickupDelay <= 0;
+	}
+
+	@Inject(method = "onPlayerCollision", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/ItemEntity;getStack()Lnet/minecraft/item/ItemStack;"), cancellable = true)
+	private void fireEntityItemPickup(PlayerEntity player, CallbackInfo ci) {
+		if (EntityEvents.onItemPickup((ItemEntity) (Object) this, player)) {
+			ci.cancel();
+		}
+	}
+
+	@ModifyConstant(method = "onPlayerCollision", constant = @Constant(intValue = 6000, ordinal = 0))
+	private int onPlayerCollideCheckLifespan(int original) {
+		return lifespan;
+	}
+
+	@Redirect(method = "onPlayerCollision", at = @At(value = "INVOKE", target = "Lnet/minecraft/entity/player/PlayerEntity;sendPickup(Lnet/minecraft/entity/Entity;I)V"))
+	private void onPlayerItemPickup(PlayerEntity entity, Entity item, int count) {
+		EntityEvents.firePlayerItemPickupEvent(entity, (ItemEntity) item, getStack().copy());
```

Comment:
I also implemented this event in #118.
`patchwork-events-entity/src/main/java/net/patchworkmc/mixin/event/entity/MixinItemEntity.java`

I created `PlayerEvents` to fire all types of PlayerEvent, so that `EntityEvents` can be less messy. I would suggest one of us should drop his `PlayerEvent.ItemPickupEvent`  implementation. Sorry for the inconvenience, I didn't know this PR before I started #118.

Something I would like to discuss is that, in 1.14.4 (The yarnforge snapshot), Forge removed the `entityIn.sendPickup(this, i);` before `if (itemstack.isEmpty()) {` but in the later version they didnt. However, I don't think firing the PlayerEvent.ItemPickupEvent before or after `entityIn.sendPickup(this, i);` make any noticable difference.